lib/codecs/src/encoding/format/cef.rs

```diff
@@ -0,0 +1,520 @@
+
+    fn encode(&mut self, event: Event, buffer: &mut BytesMut) -> Result<(), Self::Error> {
+        let log = event.into_log();
+
+        let severity: u8 = match get_log_event_value(&log, &self.severity).parse() {
+            Err(err) => {
+                return SeverityNumberTypeSnafu { error: err }
+                    .fail()
+                    .map_err(|e| e.to_string().into());
+            }
+            Ok(severity) => {
+                if severity > SEVERITY_MAX {
+                    return SeverityMaxValueSnafu {
+                        max_value: SEVERITY_MAX,
+                        actual_value: severity,
+                    }
+                    .fail()
+                    .map_err(|e| e.to_string().into());
+                };
+                severity
+            }
+        };
+
+        let name: String = escape_header(&get_log_event_value(&log, &self.name));
+        if name.len() > NAME_MAX_LENGTH {
+            return ExceededLengthSnafu {
+                field: name.clone(),
+                max_length: NAME_MAX_LENGTH,
+                actual_length: name.len(),
+            }
+            .fail()
+            .map_err(|e| e.to_string().into());
+        };
+
+        let mut formatted_extensions = Vec::with_capacity(self.extensions.len());
+        for (extension, field) in &self.extensions {
+            let value = get_log_event_value(&log, field);
+            if value.is_empty() {
+                continue;
+            }
+            let value = escape_extension(&value);
+            formatted_extensions.push(format!("{}={}", extension, value));
+        }
+
+        buffer.write_fmt(format_args!(
+            "CEF:{}|{}|{}|{}|{}|{}|{}",
+            &self.version.as_str(),
+            &self.device.vendor,
+            &self.device.product,
+            &self.device.version,
+            &self.device.event_class_id,
+            name,
+            severity,
+        ))?;
+        if !formatted_extensions.is_empty() {
+            formatted_extensions.sort();
+
+            buffer.write_char('|')?;
+            buffer.write_str(formatted_extensions.join(" ").as_str())?;
+        }
+
+        Ok(())
+    }
+}
+
+fn get_log_event_value(log: &LogEvent, field: &ConfigTargetPath) -> String {
+    match log.get(field) {
+        Some(Value::Bytes(bytes)) => String::from_utf8_lossy(bytes).to_string(),
+        Some(Value::Integer(int)) => int.to_string(),
+        Some(Value::Float(float)) => float.to_string(),
+        Some(Value::Boolean(bool)) => bool.to_string(),
+        // TODO: support other timestamp options.
+        Some(Value::Timestamp(timestamp)) => timestamp.to_rfc3339_opts(SecondsFormat::AutoSi, true),
+        Some(Value::Null) => String::from(""),
+        // Other value types: Array, Regex, Object are not supported by the CEF format.
+        Some(_) => String::from(""),
+        None => String::from(""),
+    }
+}
+
+fn escape_header(s: &str) -> String {
```

Comment:
This and the following can be combined in one:
```
fn escape_special_chars(s: &str, extra_char: char) -> String {
    s.replace('\\', r#"\\"#).replace(extra_char, &format!(r#"\{}"#, extra_char))
}
```

Reply:
Fixed, but there are now three functions 😅